src/lightning/pytorch/core/datamodule.py

```diff
@@ -223,11 +223,12 @@ def load_from_checkpoint(
             )
 
         """
-        return _load_from_checkpoint(
+        loaded = _load_from_checkpoint(
             cls,
             checkpoint_path,
             map_location=None,
             hparams_file=hparams_file,
             strict=None,
             **kwargs,
         )
+        return cast(Self, loaded)
```

Comment:
I tried to avoid this cast, but it would really complicate the `_load_from_checkpoint` typing. We would need TypeVars bounded to LM and LDM + maybe some overloads. Couldn't get it to work.

Another solution would be to inline the `_load_from_checkpoint` implementation here (and for the LDM), so that the return type is not ambiguous